polonius-engine/src/output/datafrog_opt.rs

```diff
@@ -16,33 +16,33 @@ use crate::output::liveness;
 use crate::output::Output;
 
 use datafrog::{Iteration, Relation, RelationLeaper};
-use facts::{AllFacts, Atom};
+use facts::{AllFacts, FactTypes};
 
-pub(super) fn compute<Origin: Atom, Loan: Atom, Point: Atom, Variable: Atom, MovePath: Atom>(
+pub(super) fn compute<T: FactTypes>(
     dump_enabled: bool,
-    all_facts: AllFacts<Origin, Loan, Point, Variable, MovePath>,
-) -> Output<Origin, Loan, Point, Variable, MovePath> {
+    all_facts: AllFacts<T>,
+) -> Output<T> {
     let mut result = Output::new(dump_enabled);
 
     let var_maybe_initialized_on_exit = initialization::init_var_maybe_initialized_on_exit(
-        all_facts.child,
-        all_facts.path_belongs_to_var,
-        all_facts.initialized_at,
-        all_facts.moved_out_at,
-        all_facts.path_accessed_at,
-        &all_facts.cfg_edge,
+        all_facts.child.clone(),
+        all_facts.path_belongs_to_var.clone(),
```

Comment:
Why do we now require all of these `clone` calls here?

Reply:
This might be related to the re-jiggering I am talking about: the initialisation function moves its inputs out of the struct as it's the only step using them (which is why I would have suggested splitting the struct). I'm not sure if the borrow checker isn't smart enough to see that they are only used in this function when this refactoring is applied? Either that, or something else. :)